### PR TITLE
LPS-159413 Improve form alerts when publishing

### PR DESCRIPTION
**Pull request by sandrodw3**

### CODE REVIEW

**Comment by liferay-echo** [comment]
Thank you for submitting this PR. 👏

Make sure it follows the [Echo Pull Request Template](https://gist.github.com/liferay-echo/5a4455285ff1800f9da146816d85e67d#file-pull-request-template-md) if you haven't already done so. 

Learn more about it [here](https://liferay.atlassian.net/wiki/spaces/ENGECHO/pages/2027455014/Echo+Pull+Request+Etiquette).

**Comment by sandrodw3** [comment]
ci:test:sf

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/contexts/FormValidationContext.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+const useFormErrors = () => {
+	const {errors} = useContext(FormValidationContext);
+
+	return errors;
+};
+
+const useSetFormErrors = () => {
+	const {setErrors} = useContext(FormValidationContext);
+
+	return (errors) => setErrors(errors);
```
Comment (p2kmgcl) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/contexts/FormValidationContext.js:43]: We can just return "setErrors" here :sweat_smile:

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+import {getFormValidationData} from '../utils/getFormValidationData';
+
+export function FormValidationModal({onCloseModal, onPublish}) {
+	const {observer, onClose} = useModal({
+		onClose: onCloseModal,
+	});
+
+	const formErrors = useFormErrors();
+
+	const formIds = Array.from(formErrors.keys());
```
Comment (p2kmgcl) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:32]: Maybe we should use a memo here

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+								if (!formIds[nextIndex]) {
+									onPublish();
+								}
+								else {
+									setCurrentIndex(nextIndex);
+								}
+							}}
+						>
+							{Liferay.Language.get('publish')}
```
Comment (p2kmgcl) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:91]: Is a bit weird that you have to click publish many times until there are no more forms with errors. Shouldn't we change this label to something like "check next form", group all forms together, or give some progress feedback?

**Review by p2kmgcl — COMMENTED** [review]
Very cool, I just added some small refactor suggestions :rainbow: 
Also I think we should merge the two buildLang commits.

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/contexts/FormValidationContext.js (hunk 1 above)
Comment (sandrodw3) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/contexts/FormValidationContext.js:43]: That's right 😅

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js (hunk 2 above)
Comment (sandrodw3) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:32]: Yep, you are right, we'll avoid some extra calculations

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js (hunk 3 above)
Comment (sandrodw3) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:91]: That's a good point 🥰 I'll check it with Design!

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/utils/useCheckFormsValidity.js [hunk 4]
```diff
@@ -30,9 +32,19 @@ function getFormItems(layoutData) {
...
 
-export default function useIsSomeFormIncomplete() {
+function addError(errors, formId, type) {
+	const formErrors = errors.get(formId);
+	const nextFormErrors = formErrors ? [...formErrors, type] : [type];
+
+	errors.set(formId, nextFormErrors);
+}
+
+export default function useCheckFormsValidity() {
```
[the remark this answers is not in the document]
Comment (sandrodw3) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/utils/useCheckFormsValidity.js:28]: Thanks 🥰

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+				return null;
+			}
+
+			const type = config.formTypes.find(
+				({value}) => value === item.config.classNameId
+			);
+
+			return type.label;
+		},
+		[currentIndex]
```
[the remark this answers is not in the document]
Comment (sandrodw3) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:52]: Done 😅

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+import {getFormErrorDescription} from '../utils/getFormErrorDescription';
+
+export function FormValidationModal({onCloseModal, onPublish}) {
+	const {observer, onClose} = useModal({
+		onClose: onCloseModal,
+	});
+
+	const validation = useFormValidation();
+
+	if (validation.length === 1 && validation[0].errors.length === 1) {
```
Comment (p2kmgcl) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:33]: I though validation has a `Map`, `validation[0]` does not work, or does it?

Comment (sandrodw3) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/FormValidationModal.js:33]: It's a map when I'm calculating it but I convert it to array when sending the data to the context. I'll correct the initialization in the context, that is what confused you, sorry 😅

**Comment by sandrodw3** [comment]
ci:forward

**Comment by jkappler** [comment]
@sandrodw3, @p2kmgcl, fixed conflicts and resent here: https://github.com/liferay-echo/liferay-portal/pull/9254

**Pull request closed without merging**